anvil/src/state.rs

```diff
@@ -388,15 +390,85 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
         }
     }
 
-    pub fn send_frames(&self, output: &Output) {
-        self.space.elements().for_each(|window| {
-            if self.space.outputs_for_element(window).contains(output) {
-                window.send_frame(self.start_time.elapsed().as_millis() as u32)
+    pub fn send_frames(&self, output: &Output, report: &OutputRenderReport) {
+        let now = Instant::now();
+        let throttle = Some(Duration::from_secs(1));
+        let time = self.start_time.elapsed().as_millis() as u32;
+
+        for window in self.space.elements() {
+            let outputs = self.space.outputs_for_element(window);
+
+            // If the element outputs do not contain
+            // the requested output we can exit early
+            // as all further checks for the primary output
+            // will fail as well
+            if !outputs.contains(output) {
+                continue;
             }
-        });
+
+            window.send_frame(time, |surface, states| {
+                // We do not want to send frame callbacks triggered by a
+                // non primary output
+                if !window_surface_primary_output(surface, &*outputs)
+                    .map(|primary_output| primary_output == output)
+                    .unwrap_or(false)
+                {
+                    return false;
+                }
+
+                send_throttled_frame(surface, states, report, now, throttle)
+            });
+        }
+
         let map = smithay::desktop::layer_map_for_output(output);
         for layer_surface in map.layers() {
-            layer_surface.send_frame(self.start_time.elapsed().as_millis() as u32)
+            layer_surface.send_frame(time, |surface, states| {
+                send_throttled_frame(surface, states, report, now, throttle)
+            });
+        }
+    }
+}
+
+fn send_throttled_frame(
+    surface: &WlSurface,
+    states: &smithay::wayland::compositor::SurfaceData,
+    report: &OutputRenderReport,
+    tp: Instant,
+    throttle: Option<Duration>,
+) -> bool {
+    let throttle_frame = report
+        .element_render_state(surface)
+        .map(|state| state == RenderElementState::Skipped)
+        .unwrap_or(true);
+
+    states
+        .data_map
+        .insert_if_missing_threadsafe(SurfaceFrameThrottlingState::default);
+    let send_throttled_frame = states
+        .data_map
+        .get::<SurfaceFrameThrottlingState>()
+        .unwrap()
+        .update(tp, throttle);
+
+    !throttle_frame || send_throttled_frame
```

Comment:
Again the naming, especially with the inverse logic, makes this expression hard to parse.
Something like:
`!was_skipped || frame_overdue` or something along those lines would be easier to understand.

Reply:
now uses `presented || frame_overdue`